main/src/library/Chain.flix

```diff
@@ -0,0 +1,865 @@
+/*
+ * Copyright 2021 Stephen Tetley
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+///
+/// The Chain type.
+///
+/// A chain is a list represented as an unbalanced binary tree.
+/// It supports efficient append and "snoc" - appending elements at the tail
+/// of the list.
+///
+/// Note - the constructors `Empty`, `One` and `Chain` should not be used directly.
+///
+enum Chain[t] {
+    case Empty
+    case One(t)
+    case Chain(Chain[t], Chain[t])
+}
+
+instance ToString[Chain[a]] with ToString[a] {
```

Comment:
Could I ask you to move the ToString instance after all the other instances?

I appreciate that the other instances are declared in a natural order: SemiGroup, Monoid, Functor into Applicative into Monad. 👍

Reply:
Fixed.